text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,570 @@
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent:*",
+  },
+}
+```
+
+On publish, the least suprising behavior would be for this to become `17.0.2`. This is because `17.0.2` was the resolved version actually tested on in the monorepo.
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "17.0.2",
+  },
+}
+```
+
+This leads to a discrepancy between what `workspace-consistent:*` and `workspace-consistent:^` should append semver range types to on publish. (The original specifier vs the resolved version.)
+
+The semantics may become more clear if `workspaceConsistent` configured versions must be a concrete version, but this eliminates the ability to `pnpm update` `workspaceConsistent` values.
+
+### Lockfile `workspaceConsistent` block format
+
+The proposed format is:
+
+```yaml
+workspaceConsistent:
+  jest:
+    specifier: ^27.5.1
+    dependency: 27.5.1
+
+  react:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+
+  react-dom:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+```
+
+An alternative format that represents the same information would be:
+
+```yaml
+workspaceConsistent:
+  specifiers:
+    jest: ^27.5.1
+    react: ^17.0.2
+    react-dom: ^17.0.2
+  dependencies:
+    jest: 27.5.1
+    react: 17.0.2
+    react-dom: 17.0.2
+```
+
+The alternative format better matches the existing shape of `importers`. However the proposed format is more resilient to merge conflicts. Separate commits changing the version of `react` and `jest` would merge conflict with the alternative format, while the proposed format would merge those changes cleanly.
+
+## Extensions
+
+### Transitive Consistency
+
+Multiple versions of a package can still end up in the final product due to versions declared as a dependency of a dependency.
+
+It would be useful to declare two extra fields to solve the duplicate dependency problem completely.
+
+```json5
+{
+  "name": "example-monorepo",
+  "private": true,
+  "pnpm": {
+    "workspaceConsistent": {
+      "groups": {
+        "default": {
+          "@types/react": "^17.0.43",
+          "react": "^17.0.2",
+          "react-dom": "^17.0.2",
+        }
+      },
+      "enforceConsistencyTransitively": true,
+      "allowTransitiveMismatches": ["@types/react"]
+    }
+  }
+}
+```
+
+`enforceConsistencyTransitively` will error if there is ever more than one item in the `packages` block of `pnpm-lock.yaml` for a dependency, even duplicates on the same version but appearing multiple times due to peer dependencies.
+
+The `allowTransitiveMismatches` option will provide an escape hatch in case specific dependencies are acceptable. In the case above, duplicate `@types/react` dependencies may be acceptable since it's usually under `devDependencies` and package authors may declare a semver range without any overlap on the monorepo workspace consistent range.
```

Comment:
In a TypeScript repository, `@types/react` is a particularly bad example for allowing duplicates, since duplicate versions of typings tend to cause confusingly worded type checking errors when types from one version fail to assign to the types in another version, for example if a library exports some react components that are then consumed in another project, they need to have the exact same version of `@types/react`.

Reply:
You're right. This was a bad example. The scenario you described is one of the main motivations for this RFC. Switched this to `object-assign` since that's an easy example small runtime example and `react` depends on it.